### PR TITLE
Mesh object generation from parametric surfaces

### PR DESCRIPTION
**Pull request by docwhite**
Built in primitives implemented are torus, sphere and grid. Issue #700

### CODE REVIEW

**Comment by Narann** [comment]
So cool! Thanks!

**Comment by dictoon** [comment]
Awesome, thanks!

File: sandbox/tests/test scenes/primitive mesh/01 - sphere.appleseed [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+            </parameters>
+            <parameters name="uniform_pixel_renderer">
+                <parameter name="decorrelate_pixels" value="true" />
+                <parameter name="force_antialiasing" value="false" />
+                <parameter name="samples" value="16" />
+            </parameters>
+        </configuration>
+        <configuration name="interactive" base="base_interactive">
+            <parameter name="lighting_engine" value="pt" />
+            <parameter name="rendering_threads" value="1" />
```
Comment (dictoon) [inline sandbox/tests/test scenes/primitive mesh/01 - sphere.appleseed]: It's best not to force a specific number of rendering threads in test scenes. You could either remove this line or remove the override in the Interactive Mode rendering settings in appleseed.studio.

File: sandbox/tests/test scenes/primitive mesh/02 - plane.appleseed [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+            </parameters>
+            <parameters name="uniform_pixel_renderer">
+                <parameter name="decorrelate_pixels" value="true" />
+                <parameter name="force_antialiasing" value="false" />
+                <parameter name="samples" value="16" />
+            </parameters>
+        </configuration>
+        <configuration name="interactive" base="base_interactive">
+            <parameter name="lighting_engine" value="pt" />
+            <parameter name="rendering_threads" value="1" />
```
Comment (dictoon) [inline sandbox/tests/test scenes/primitive mesh/02 - plane.appleseed]: Same thing here and for all remaining test scenes.

File: src/appleseed/renderer/modeling/object/meshobjectprimitives.cpp [hunk 3]
```diff
@@ -0,0 +1,251 @@
+
+//
+// This source file is part of appleseed.
+// Visit http://appleseedhq.net/ for additional information and resources.
+//
+// This software is released under the MIT license.
+//
+// Copyright (c) 2016 Ramon Blanquer
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/object/meshobjectprimitives.cpp]: The copyright notice should be

    // Copyright (c) 2016 Ramon Blanquer, The appleseedhq Organization

File: src/appleseed/renderer/modeling/object/meshobjectprimitives.cpp [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+
+using namespace foundation;
+using namespace std;
+
+namespace renderer
+{
+
+class ParametricSphere
+{
+public:
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/object/meshobjectprimitives.cpp]: Indent visibility keywords with two spaces:

    class ParametricSphere
    {
      public:
        explicit ParametricSphere(...)

File: src/appleseed/renderer/modeling/object/meshobjectprimitives.cpp [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+    explicit ParametricSphere(const double radius)
+      : m_radius(radius)
+    {
+    }
+
+    GVector3 evaluate(const double u, const double v) const
+    {
+        const double theta = 2.0 * Pi * u;
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/object/meshobjectprimitives.cpp]: You can use the `TwoPi` constant to save one multiplication here (this won't necessarily be done by the compiler, depending on the floating point mode).

File: src/appleseed/renderer/modeling/object/meshobjectprimitives.cpp [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+        const double phi = Pi * v;
+
+        return GVector3(
+            m_radius * cos(theta) * sin(phi),
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/object/meshobjectprimitives.cpp]: Not that that's necessarily a hotspot but you could compute `sin(phi)` once.

File: src/appleseed/renderer/modeling/object/meshobjectprimitives.cpp [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+           -m_radius * cos(phi),
+            m_radius * sin(theta) * sin(phi)
+        );
+    }
+
+    const double m_radius;
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/object/meshobjectprimitives.cpp]: Make `m_radius` private?

File: src/appleseed/renderer/modeling/object/meshobjectprimitives.cpp [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+            const GVector3 p = surface.evaluate(u, v);
+
+            // Compute u tangent using forward differencing.
+            GVector3 dpdu = surface.evaluate(u + h, v) - p;
+            const double dpdu_norm = norm(dpdu);
+
+            if (dpdu_norm == 0.0)
+            {
+                // If the u tangent is zero (surface pole), approximate it using a nearby point.
+                dpdu = normalize(surface.evaluate(u + h, v + h) - p);
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/object/meshobjectprimitives.cpp]: What happens when `surface.evaluate(u + h, v + h) - p` is a null vector, i.e. when `surface.evaluate(u + h, v + h)` is equal to `surface.evaluate(u, v)`?

File: src/appleseed/renderer/modeling/object/meshobjectprimitives.cpp [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+            (m_major_radius + m_minor_radius * cos(phi)) * cos(theta),
+             m_minor_radius * sin(phi),
+            (m_major_radius + m_minor_radius * cos(phi)) * sin(theta));
+    }
+
+    const double m_major_radius;
+    const double m_minor_radius;
+};
+
+template<typename ParametricSurface>
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/object/meshobjectprimitives.cpp]: Add space before `<`:

    template <typename ParametricSurface>

File: src/appleseed/renderer/modeling/object/meshobjectprimitives.cpp [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+{
+    create_vertices(mesh, surface, resolution_u, resolution_v);
+    create_triangles(mesh, resolution_u, resolution_v);
+    mesh.push_material_slot("default");
+}
+
+auto_release_ptr<MeshObject> create_primitive_mesh(const char* name, const ParamArray& params)
+{
+    auto_release_ptr<MeshObject> mesh = MeshObjectFactory::create(name, params);
+    const size_t resolution_u  = params.get_optional<size_t>("resolution_u" , 32);
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/object/meshobjectprimitives.cpp]: Remove extra space before comma.

**Comment by dictoon** [comment]
Exceptional code quality! A few minor things to fix and it's good to merge.

Thanks!

File: src/appleseed/renderer/modeling/object/meshobjectprimitives.cpp (hunk 8 above)
Comment (docwhite) [inline src/appleseed/renderer/modeling/object/meshobjectprimitives.cpp]: I added validation on the parameters and I think it cannot happen anymore with the primitives implemented.

**Comment by docwhite** [comment]
Cheers @dictoon @Narann :smiley: !!

**Comment by dictoon** [comment]
Perfect, thank you!

**Pull request merged**